### PR TITLE
Federated Ingress: unify UID's across Cluster Ingress Controllers

### LINKED ISSUE #31180: Federated Ingress Controller needs to unify UID's across Cluster Ingress Controllers
See "UID Assignment" in https://gist.github.com/bprashanth/52648b2a0b6a5b637f843e7efb2abc97

PR in the works.

### PR DESCRIPTION
**Pull request by ghost**
Fixes #31180

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31600)

### CODE REVIEW

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 1]
```diff
@@ -240,6 +346,131 @@ func (ic *IngressController) reconcileIngressesOnClusterChange() {
...
+  updateClusterIngressUIDToMasters takes the ingress UID annotation on the master cluster and applies it to cluster.
+  If there is no master cluster, then fallbackUID is used (and hence this cluster becomes the master).
+*/
+func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federation_api.Cluster, fallbackUID string) {
+	masterCluster, masterUID, err := ic.getMasterCluster()
+	if err != nil {
+		if masterCluster.Name != cluster.Name { // If we are the master, no update needed
+			cluster.ObjectMeta.Annotations[uidAnnotationKey] = masterUID
+			if _, err = ic.federatedApiClient.Federation().Clusters().Update(cluster); err != nil {
+				glog.Errorf("Failed to add master ingress UID annotation (%q = %q) from master cluster %q to cluster %q, will try again later: %v", uidAnnotationKey, masterUID, masterCluster.Name, cluster.Name, err)
```
Comment (bprashanth) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: so we'll change the uid if a GET fails?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 2]
```diff
@@ -158,6 +215,25 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
+	// Federated configmap updater along with Create/Update/Delete operations.  Only Update should ever be called.
+	ic.federatedConfigMapUpdater = util.NewFederatedUpdater(ic.configMapFederatedInformer,
+		func(client kube_release_1_4.Interface, obj pkg_runtime.Object) error {
+			configMap := obj.(*api.ConfigMap)
+			configMapName := types.NamespacedName{Name: configMap.Name, Namespace: configMap.Namespace}
+			return fmt.Errorf("Internal error: Incorrectly attempting to create ConfigMap: %q", configMapName)
+		},
+		func(client kube_release_1_4.Interface, obj pkg_runtime.Object) error {
+			configMap := obj.(*v1.ConfigMap)
+			glog.V(4).Infof("Attempting to update ConfigMap: %v", configMap)
```
Comment (bprashanth) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: changing the uid once you have working ingresses is not easy. It requires recreation of cloud resources and hance downtime. Typically users should not want to do this, the recommended way to handle it is to delete ingresses, change uid, recreate ingress (https://github.com/kubernetes/contrib/blob/master/ingress/controllers/gce/BETA_LIMITATIONS.md#changing-the-cluster-uid). It would be even worse if one cluster ended up with a different uid as it was creating ingress.

Can you complain loudly anytime we actually update it (event against master, Warningf logging)?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 3]
```diff
@@ -48,14 +55,25 @@ type IngressController struct {
...
 
 	// Contains ingresses present in members of federation.
 	ingressFederatedInformer util.FederatedInformer
-	// For updating members of federation.
-	federatedUpdater util.FederatedUpdater
+	// Contains ingress controller configmaps present in members of federation.
+	configMapFederatedInformer util.FederatedInformer
+	// For updating ingresses in members of federation.
+	federatedIngressUpdater util.FederatedUpdater
+	// For updating ingresses in members of federation.
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: sounds more like config map updater

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 4]
```diff
@@ -78,14 +99,19 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
 		clusterAvailableDelay: time.Second * 20,
 		smallDelay:            time.Second * 3,
 		updateTimeout:         time.Second * 30,
 		ingressBackoff:        flowcontrol.NewBackOff(5*time.Second, time.Minute),
+		configMapBackoff:      flowcontrol.NewBackOff(5*time.Second, time.Minute),
 	}
 
 	// Build deliverers for triggering reconcilations.
 	ic.ingressDeliverer = util.NewDelayingDeliverer()
+	/* TODO Remove
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Please remove, here and below.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 5]
```diff
@@ -132,14 +158,45 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
+						return targetClient.Core().ConfigMaps(api.NamespaceSystem).Watch(options) // TODO: As above
+					},
+				},
+				&api.ConfigMap{},
+				controller.NoResyncPeriodFunc(),
+				// Trigger reconcilation whenever the ingress controller's configmap in a federated cluster is changed. In most cases it
+				// would be just confirmation that the configmap for the ingress controller is correct.
+				util.NewTriggerOnAllChanges(
+					func(obj pkg_runtime.Object) {
+						ic.reconcileConfigMapForCluster(cluster.Name)
```
Comment (mwielgus) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Reconcilation should happen on a separate thread. Reconciliating here you will block the whole informer.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Whoops, yes.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 4 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yup, will do.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 5 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: ok

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 2 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yes, agreed.  Will do, and add appropriate release notes.  It would actually be pretty easy to refuse to do this here if any Ingresses already exist in the cluster.  See also https://github.com/kubernetes/contrib/issues/1656 for the better plan for beta and GA, as this is clearly a fairly significant limitation of Federated Ingress alpha.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 1 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: No, once a master cluster has been chosen once, it's UID is applied to all other clusters (in the federation etcd, and thence to the cluster's ingress controller configmaps), and if any one of them becomes the master in future, the UID will remain constant.  This only deals with the case where none of the clusters (in the federation etcd) know what the UID is (e.g. when only one cluster is registered), in which case the one from that (now master) cluster is used.

It would be better to persist this in one central place rather than in all of the registered clusters, but place does not exist yet.  I will add an issue to track improving this.

**Comment by quinton-hoole** [comment]
Added commit to address all but one reviewer comment.

**Comment by quinton-hoole** [comment]
ok to test

**Comment by quinton-hoole** [comment]
@mwielgus Would appreciate a final review of this please.  All unit tests pass now.  I can parallelize the updates better in a followup PR, but I think we should merge this in the mean time.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 6]
```diff
@@ -68,35 +91,41 @@ type IngressController struct {
...
 	ingressReviewDelay    time.Duration
+	configMapReviewDelay  time.Duration
 	clusterAvailableDelay time.Duration
 	smallDelay            time.Duration
 	updateTimeout         time.Duration
 }
 
 // NewIngressController returns a new ingress controller
 func NewIngressController(client federation_release_1_4.Interface) *IngressController {
+	glog.V(4).Infof("->NewIngressController V(4)")
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: I realize that this function is way too long and in need of refactoring - will do in a followup PR.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 7]
```diff
@@ -218,40 +338,240 @@ func (ic *IngressController) deliverIngress(ingress types.NamespacedName, delay
...
+			}
+		} else {
+			glog.Errorf("No master cluster exists, and fallbackUID for cluster %q is invalid (%q).  This probably means that no clusters have an ingress controller configmap with key %q.  Federated Ingress currently supports clusters running Google Loadbalancer Controller (\"GLBC\")", cluster.Name, fallbackUID)
+		}
 	}
 }
 
 func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
-	glog.V(4).Infof("Reconciling ingress %q", ingress)
+	glog.V(4).Infof("Reconciling ingress %q for all clusters", ingress)
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: I realize that this function is way too long and in need of refactoring - will do in a followup PR.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 8]
```diff
@@ -218,40 +338,240 @@ func (ic *IngressController) deliverIngress(ingress types.NamespacedName, delay
...
+		}
+	}
+	return nil, "", fmt.Errorf("Failed to find master cluster with annotation %q", uidAnnotationKey)
+}
+
+/*
+  updateClusterIngressUIDToMasters takes the ingress UID annotation on the master cluster and applies it to cluster.
+  If there is no master cluster, then fallbackUID is used (and hence this cluster becomes the master).
+*/
+func (ic *IngressController) updateClusterIngressUIDToMasters(cluster *federation_api.Cluster, fallbackUID string) {
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: This ideally needs to be done in parallel via a worker queue.  Will do in a followup PR.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 9]
```diff
@@ -307,71 +652,92 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
...
+					lbstatus, ok := lbstatusObj.(*v1.LoadBalancerStatus)
+					if lbErr != nil || !ok {
+						glog.Errorf("Internal error: Failed to clone LoadBalancerStatus of %q in cluster %q while attempting to update master loadbalancer ingress status, will try again later. error: %v, Object to be cloned: %v", ingress, cluster.Name, lbErr, lbstatusObj)
+						ic.deliverIngress(ingress, ic.ingressReviewDelay, true)
+						return
 					}
+					baseIngress.Status.LoadBalancer = *lbstatus
+				}
+				glog.V(4).Infof("Attempting to update base federated ingress: %v", baseIngress)
+				if _, err = ic.federatedApiClient.Extensions().Ingresses(baseIngress.Namespace).Update(baseIngress); err != nil {
```
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Ideally this needs to be done in parallel via a worker queue.  Will do in a followup PR.

**Comment by nikhiljindal** [comment]
Jenkins verify failed with `federation/pkg/federation-controller/ingress/ingress_controller.go:568: missing argument for Errorf("%q"): format reads arg 3, have only 2 args`

**Comment by quinton-hoole** [comment]
Thanks @nikhiljindal .  Fixed.  I'll keep an eye on the jenkins test results and fix whatever comes up. Unfortunately the verification scripts hang on my Mac :-(

**Comment by quinton-hoole** [comment]
All tests passed. Only awaiting LGTM now.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 10]
```diff
@@ -40,24 +44,43 @@ import (
 )
 
 const (
-	allClustersKey        = "ALL_CLUSTERS"
-	staticIPAnnotationKey = "ingress.kubernetes.io/static-ip" // TODO: Get this directly from the Kubernetes Ingress Controller constant
+	// Special cluster name which denotes all clusters - only used internally.  It's not a valid cluster name, so effectively reserved.
+	allClustersKey          = ".ALL_CLUSTERS"
+	staticIPNameKeyWritable = "kubernetes.io/ingress.global-static-ip-name" // The writable annotation on Ingress to tell the controller to use a specific, named, static IP
+	// TODO: Get the above directly from the Kubernetes Ingress Controller constant - but it's in a separate repo
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: I assume this TODO is not only for the above, but the contants below as well.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 11]
```diff
@@ -68,35 +91,41 @@ type IngressController struct {
...
 func NewIngressController(client federation_release_1_4.Interface) *IngressController {
+	glog.V(4).Infof("->NewIngressController V(4)")
 	broadcaster := record.NewBroadcaster()
 	broadcaster.StartRecordingToSink(eventsink.NewFederatedEventSink(client))
 	recorder := broadcaster.NewRecorder(api.EventSource{Component: "federated-ingress-controller"})
-
 	ic := &IngressController{
 		federatedApiClient:    client,
-		ingressReviewDelay:    time.Second * 10,
+		ingressReviewDelay:    time.Minute * 3,
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Why are they so big (3 minutes and 6 minutes)? Why not in seconds (as other delays here).

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 12]
```diff
@@ -142,24 +171,69 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
-	ic.federatedUpdater = util.NewFederatedUpdater(ic.ingressFederatedInformer,
+	// Federated informer on configmaps for ingress controllers in members of the federation.
+	ic.configMapFederatedInformer = util.NewFederatedInformer(
+		client,
+		func(cluster *federation_api.Cluster, targetClient kube_release_1_4.Interface) (cache.Store, framework.ControllerInterface) {
+			glog.V(4).Infof("Returning new informer for cluster %q", cluster.Name)
+			return framework.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						if targetClient == nil {
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Why is this only in ListFunc and not in WatchFunc?
Is it better to move it to the top, where function definition starts (line 104)

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 13]
```diff
@@ -142,24 +171,69 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
+							glog.Errorf("targetClient is nil")
+						}
+						return targetClient.Core().ConfigMaps(uidConfigMapNamespace).List(options) // we only want to list one by name - unfortunately Kubernetes don't have a selector for that.
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: What do you mean by list one by name?
Do you want Get?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 14]
```diff
@@ -142,24 +171,69 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
+	// Federated ingress updater along with Create/Update/Delete operations.
+	ic.federatedIngressUpdater = util.NewFederatedUpdater(ic.ingressFederatedInformer,
 		func(client kube_release_1_4.Interface, obj pkg_runtime.Object) error {
 			ingress := obj.(*extensions_v1beta1.Ingress)
 			glog.V(4).Infof("Attempting to create Ingress: %v", ingress)
 			_, err := client.Extensions().Ingresses(ingress.Namespace).Create(ingress)
+			if err != nil {
+				glog.Errorf("Error creating ingress %q: %v", types.NamespacedName{Name: ingress.Name, Namespace: ingress.Namespace}, err)
+			} else {
+				glog.V(4).Infof("Successfully created ingress %q", types.NamespacedName{Name: ingress.Name, Namespace: ingress.Namespace})
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: future improvement: This seems like a good place to generate an event.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 15]
```diff
@@ -168,6 +242,35 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
 			return err
 		})
+
+	// Federated configmap updater along with Create/Update/Delete operations.  Only Update should ever be called.
+	ic.federatedConfigMapUpdater = util.NewFederatedUpdater(ic.configMapFederatedInformer,
+		func(client kube_release_1_4.Interface, obj pkg_runtime.Object) error {
+			configMap := obj.(*v1.ConfigMap)
+			configMapName := types.NamespacedName{Name: configMap.Name, Namespace: configMap.Namespace}
+			glog.Errorf("Internal error: Incorrectly attempting to create ConfigMap: %q", configMapName)
+			_, err := client.Core().ConfigMaps(configMap.Namespace).Create(configMap)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Do not create if this is unexpected?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 16]
```diff
@@ -168,6 +242,35 @@ func NewIngressController(client federation_release_1_4.Interface) *IngressContr
...
+		},
+		func(client kube_release_1_4.Interface, obj pkg_runtime.Object) error {
+			configMap := obj.(*v1.ConfigMap)
+			configMapName := types.NamespacedName{Name: configMap.Name, Namespace: configMap.Namespace}
+			glog.V(4).Infof("Attempting to update ConfigMap: %v", configMap)
+			_, err := client.Core().ConfigMaps(configMap.Namespace).Update(configMap)
+			if err == nil {
+				glog.V(4).Infof("Successfully updated ConfigMap %q", configMapName)
+			} else {
+				glog.V(4).Infof("Failed to updated ConfigMap %q: %v", configMapName, err)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: updated -> update

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 17]
```diff
@@ -176,25 +279,42 @@ func (ic *IngressController) Run(stopChan <-chan struct{}) {
...
+			glog.V(4).Infof("Cluster change delivered for cluster %q, reconciling configmap and ingress for that cluster", clusterName)
+		} else {
+			glog.V(4).Infof("Cluster change delivered for all clusters, reconciling configmaps and ingresses for all clusters")
+		}
+		ic.reconcileIngressesOnClusterChange(clusterName)
+		ic.reconcileConfigMapForCluster(clusterName)
+	})
+	ic.configMapDeliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		clusterName := item.Key
+		glog.V(4).Infof("ConfigMap change delivered for cluster %q, reconciling configmap for that cluster", clusterName)
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Do the same "for all clusters"/"cluster %q" thing as above?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 18]
```diff
@@ -307,71 +652,92 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
...
+			} else {
+				glog.V(4).Infof(logStr, "Not transferring")
 			}
-			// Update existing ingress, if needed.
-			if !util.ObjectMetaIsEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) ||
-				!reflect.DeepEqual(desiredIngress.Spec, clusterIngress.Spec) {
-				// TODO: In some cases Ingress controllers in the clusters add annotations, so we ideally need to exclude those from
-				// the equivalence comparison to cut down on unnecessary updates.
+			// Update existing cluster ingress, if needed.
+			if util.ObjectMetaIsEquivalent(desiredIngress.ObjectMeta, clusterIngress.ObjectMeta) &&
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: sigh, there is also ObjectMetaEquivalent: https://github.com/kubernetes/kubernetes/blob/069929bbfd13c4c598c3657a8868888ade9e0b55/federation/pkg/federation-controller/util/meta.go#L40 :)

**Comment by nikhiljindal** [comment]
This is a lot of code!
I do not claim to understand all of it, but have added a few comments.

**Comment by nikhiljindal** [comment]
Most of them are clarifying questions. Feel free to self apply the LGTM label

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 10 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yes, agreed.  Will reword.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 11 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: In theory they're only a backup in case there are bugs in the watch code.  But re-reviewing my code, those constants are used as retries on failure in a few places.  I'll put them back to their original values for now, until I've replaced those retry loops with exponential backoffs.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 12 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: It was only used for debugging, but yes, I should and will make it consistent.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 13 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yes, I'll do that.  I was initially struggling with incompatible return types, but that's easy to fix.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 14 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Agreed.  Events need to be added throughout.  I'll do all of that in a separate PR.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 15 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Yes agreed, I originally had it that way, but changed it to this warning to track down a bug.  Will switch it back (in a separate PR).

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 16 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Good catch.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 17 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: No, this is somewhat by design.  ReplicaSets and Ingresses need to be reconciled across all clusters when the cluster set changes, while for ConfigMaps it's a simple matter of reconciling for the one changed cluster.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 18 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Nice catch.  Yes, will consolidate.  Marcin and I debugged this in parallel, and his fix merged first :-)

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 10 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 11 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 12 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done.  And the if statement needs to be evaluated in the closure, not in the parent function.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 13 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Will do in a separate PR.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 17 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Ignore the above.  I was talking rubbish.   Fixed.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 18 above)
Comment (quinton-hoole) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Done.  Removed util.ObjectMetaIsEquivalent()

**Comment by quinton-hoole** [comment]
Addressed all review comments.  Applying LGTM.

**Comment by eparis** [comment]
This is unable to be cleanly and/or incredibly easily cherry-picked to the release-1.4 branch.
@quinton-hoole can you please open a cherry-pick to 1.4 yourself?

**Comment by quinton-hoole** [comment]
@eparis Sure, happy to do that. Is there a specific process you'd like me to follow, or should I just submit a separate PR to be merged into the 1.4 branch.

**Comment by eparis** [comment]
thank you @quinton-hoole !!!

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/cherry-picks.md#how-do-cherrypick-candidates-make-it-to-the-release-branch

INDIVIDUAL CHERRYPICKS, instead of batch.

**Comment by quinton-hoole** [comment]
@eparis Done!  #32623
